sql/2020/04_Fonts/04_06.font_display_with_fcp.sql

```diff
@@ -0,0 +1,71 @@
+#standardSQL
+#font_display_with_fcp
+CREATE TEMPORARY FUNCTION getFontDisplay(css STRING)
+RETURNS ARRAY < STRING > LANGUAGE js AS '''
+try {
+    var reduceValues = (values, rule) => {
+        if ('rules' in rule) {
+            return rule.rules.reduce(reduceValues, values);
+        }
+        if (!('declarations' in rule)) {
+            return values;
+        }
+        if (rule.type != 'font-face') {
+            return values;
+        }
+        rule.declarations.forEach(d => {
+            if (d.property.toLowerCase() == 'font-display') {
+                values.push(d.value);
+            }
+        });
+        return values;
+    };
+    var $ = JSON.parse(css);
+    return $.stylesheet.rules.reduce(reduceValues, []);
+} catch (e) {
+    return [];
+}
+''';
+
+SELECT
+  client,
+  font_display,
+  COUNT(DISTINCT page) AS freq_display,
+  total_page,
+  ROUND(COUNT(DISTINCT page) * 100 / total_page, 2) AS pct_display,
+  ROUND(COUNTIF(fast_fcp>=0.75)*100/COUNT(0),0) AS pct_fast_fcp_display,
+  ROUND(COUNTIF(NOT(slow_fcp >=0.25)
+    AND NOT(fast_fcp>=0.75))*100/COUNT(0),0) AS pct_mode_fcp_display,
+  ROUND(COUNTIF(slow_fcp>=0.25)*100/COUNT(0),0) AS pct_slow_fcp_display,
+FROM
+  `httparchive.almanac.parsed_css`,
+  UNNEST(getFontDisplay(css)) AS font_display  
```

Comment:
This will create one row for every `font-display` on a page, which will count the same page's FCP multiple times when grouping by the `font_display` value.

I think the way to address this would be to deduplicate values per page before aggregating.
```suggestion
FROM (
  SELECT
    DISTINCT client,
    page,
    font_display
  FROM
    `httparchive.almanac.parsed_css`,
    UNNEST(getFontDisplay(css)) AS font_display
  WHERE
    date = '2020-08-01')
```

Reply:
Thank you to give the warm explanation of `UNNEST`.
I'm thinking that shall we use `distinct page` as well to deduplicate pages. Does it make sense?
Moreover, this metric may probably aim to compare which use or not use font-dispaly.
however, I couldn't find a way on how to query those not using.
May I have your thoughts? Thank you.

Reply:
> I'm thinking that shall we use distinct page as well to deduplicate pages. Does it make sense?

The `SELECT DISTINCT` will ensure that client, page, and font_display are all unique. It would probably be clearer if I wrote it like that rather than `DISTINCT client`.

> Moreover, this metric may probably aim to compare which use or not use font-dispaly. however, I couldn't find a way on how to query those not using.

In `getFontDisplay` rather than returning an empty array, we could return `[null]`. We'd have to do this after the `reduce` function and in the `catch` block. That should give us a `NULL` value to group by in the aggregation, representing pages without `font-display`.

Reply:
I'm not sure if it would work, but another idea is to change the type of JOIN between parsed_css and the UNNEST. Right now they're joined using the comma operator, which is an implicit CROSS JOIN. Maybe using an explicit LEFT JOIN would preserve pages with no `font_display` value and they'd be `NULL` by default.

Let me know if that doesn't make sense and I can try to make a proof-of-concept query.

Reply:
No more words could be said than thank you to the warm help. 
The cool way you mentioned actually did a wonderful job to query null value using `LEFT JOIN UNNEST`. 
I'll try to apply this to 04_7/8/9 which all have the same goal to compare using or not using property.